File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []
        
        num = {
            "2": "abc", "3": "def", "4": "ghi", "5": "jkl",
            "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"
        }
        
        # base case: only one digit left
        if len(digits) == 1:
            return list(num[digits[0]])
        
        # recursive step
        smaller_combinations = self.letterCombinations(digits[1:])
        first_digit_letters = num[digits[0]]
        
        ans = []
        for letter in first_digit_letters:           # pick a letter for first digit
            for combo in smaller_combinations:       # extend with all smaller combos
                ans.append(letter + combo)
        
        return ans

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public void f(String digits,int index,HashMap<Character,String> mp,StringBuilder temp,
    List<String> ans){
        if(index>=digits.length()) {
            ans.add(temp.toString());
            return;
        }
        int len = mp.get(digits.charAt(index)).length();
        for(int i=0;i<len;i++){
            temp.append(mp.get(digits.charAt(index)).charAt(i));
            f(digits,index+1,mp,temp,ans);
            temp.deleteCharAt(temp.length() - 1);
        }
    }
    public List<String> letterCombinations(String digits) {
        HashMap<Character,String> mp = new HashMap<>();
        StringBuilder temp = new StringBuilder();
        List<String> ans = new ArrayList<>();
        mp.put('2',"abc");
        mp.put('3',"def");
        mp.put('4',"ghi");
        mp.put('5',"jkl");
        mp.put('6',"mno");
        mp.put('7',"pqrs");
        mp.put('8',"tuv");
        mp.put('9',"wxyz");
        f(digits,0,mp,temp,ans);
        return ans;
    }
}